packages/arb-validator-core/gobridge/messagesChallenge.go

```diff
@@ -0,0 +1,391 @@
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(msgHash) {
+		return errors.New("Bisect Incorrect previous state msgHash")
+	}
+
+	hashes := make([][32]byte, 0, bisectionCount)
+	hashes = append(hashes, valprotocol.MessageChallengeDataHash(
+		chainHashes[0],
+		chainHashes[1],
+		segmentHashes[0],
+		segmentHashes[1],
+		new(big.Int).Add(new(big.Int).Div(chainLength, big.NewInt(int64(bisectionCount))), new(big.Int).Mod(chainLength, big.NewInt(int64(bisectionCount)))),
+	))
+	for i := 1; i < bisectionCount; i++ {
+		hashes = append(hashes, valprotocol.MessageChallengeDataHash(
+			chainHashes[i],
+			chainHashes[i+1],
+			segmentHashes[i],
+			segmentHashes[i+1],
+			new(big.Int).Div(chainLength, big.NewInt(int64(bisectionCount)))))
+	}
+
+	c.commitToSegment(hashes)
+	c.asserterResponded()
+
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.MessagesBisectionEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+			ChainHashes:   chainHashes,
+			SegmentHashes: segmentHashes,
+			TotalLength:   chainLength,
+			Deadline:      c.client.GoEthClient.challenges[c.contractAddress].deadline,
+		},
+	})
+
+	return nil
+}
+
+func (c *messagesChallenge) OneStepProofTransactionMessage(
+	ctx context.Context,
+	lowerHashA common.Hash,
+	lowerHashB common.Hash,
+	msg message.DeliveredTransaction,
+) error {
+	messageHash := msg.CommitmentHash()
+	arbMessageHash := message.DeliveredValue(msg).Hash()
+
+	// oneStepProof
+	if !c.challenge.challengeData.challengerDataHash.Equals(hashing.SoliditySHA3(
+		lowerHashA,
+		lowerHashB,
+		messageHash,
+		arbMessageHash,
+	)) {
+		return errors.New("OneStepProofTransactionMessage Incorrect previous state")
+	}
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.OneStepProofEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+		},
+	})
+	// TODO: challenge resolution
+	// resolveChallengeAsserterWon
+
+	return nil
+}
+
+func (c *messagesChallenge) OneStepProofEthMessage(
+	ctx context.Context,
+	lowerHashA common.Hash,
+	lowerHashB common.Hash,
+	msg message.DeliveredEth,
+) error {
+	messageHash := msg.CommitmentHash()
+
+	msgType := msg.AsValue()
+	ethMsg, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewIntValue(msg.BlockNum.AsInt()),
+		value.NewIntValue(msg.MessageNum),
+		msgType,
+	})
+	ethMsgHash := ethMsg.Hash()
+
+	msgs, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewHashOnlyValue(lowerHashB, 32),
+		value.NewHashOnlyValue(ethMsgHash, 32),
+	})
+	matchHash := valprotocol.MessageChallengeDataHash(
+		lowerHashA,
+		hashing.SoliditySHA3(
+			hashing.Bytes32(lowerHashA),
+			hashing.Bytes32(messageHash)),
+		lowerHashB,
+		msgs.Hash(),
+		big.NewInt(1),
+	)
+
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(matchHash) {
```

Comment:
Why does this get the challenge data through the `GoEthClient` rather than fetching it from this class itself. Seems like needless indirection.

Reply:
cleaned up